### ISSUE TITLE
How to add features and limits for Compat after Core ships?

### ISSUE THREAD

**Issue by kainino0x** [gpuweb/gpuweb#4987]
**This discussion started on #4983. Please read the discussion there first!**

For example:

- Before shipping compat 
  - Limits that we discover but don't want to hardcode to 0 at the start
  - Features that we discover and don't want to hard-disallow at the start
- After shipping compat
  - Hardcoded limits that we want to upgrade to configurable limits
  - Hard-disallowed feature functionality that we want to upgrade to configurable features

The problems are:

- Feature detection in compat-able browsers - solved by adapter.limits/features
- Feature detection in compat-less browsers
  - Can we expect of compat-less browsers that they will ship new Core features/limits that were added *for the benefit of Compat*?

**Comment by kainino0x** [comment]
Starting with Re: https://github.com/gpuweb/gpuweb/pull/4983#issuecomment-2489544293 @mwyrzykowski 

> I think we should make the features and limits which are only needed in compat mode only cause errors if not requested in compat mode.

I'm having trouble understanding this sentence. If still needed after my wall of text, could you rephrase or give an example?

I think the smoothest way for Compat to exist is for Core to require Compat-less implementations to do a few things to make the differences smaller:
- Accept `featureLevel: "compatibility"` and ignore it
- Both expose and accept a few new limits (but no handling is required outside of initialization time)
- Expose, accept, and enable-by-default a few new feature names (but ^)

HOWEVER...

> If we don't do that and we discover a feature is needed for compatibility post v1 of the core WebGPU specification, we now potentially break all UAs by adding a feature to compat mode until the UA supports the new feature.

(I had forgotten about this, so thank you for noticing)

This very much is expected to happen, not just if we find new things for Compat that we had missed so far, but also whenever we need to add granularity to Compat so that we can expose the capabilities of devices that are _more capable_ than base Compat but _less capable_ than base Core.

So we need to design for it. (I swear I had something written up somewhere about how this should work, but I can't find it. And now thinking about it I'm not sure I did fully think it through.) Basically I think there are two options:

- Like with any other feature, when a new feature comes out, developers have to be careful about the browser support before they rely on it. Firstly, they should check the adapter for support before attempting to enable features.
  - *If* they don't, then requestDevice will fail sometimes (and they'll hit their error/fallback path). It will fail on older version of Chrome, and it will fail on older versions of Safari. Because of this Safari will need to add the feature to GPUSupportedFeatures/requiredFeatures even though it doesn't do anything.
  - And because these are new features that enabled by default in core, applications which use these features by targeting upgrades to core will essentially have to check for `"the-new-feature"` _or_ that they got Core (i.e. `"everything-from-compat-to-core"`, or detect by trying requestAdapter(core) first, then falling back to requestAdapter(compat))
    - Maybe this means we should urgently get `"everything-from-compat-to-core"` (bikeshedding tbd) landed in Core so that implementations can implement it ASAP - similar to my push for `featureLevel: "compatibility"`.
- We come up with some new mechanism for these features and limits. I'll sketch an example - maybe there is a better design I haven't thought of, but I think it will be really complicated and still be a very incomplete solution.
  - Features always start with `compatibility-` and are ignored by Safari's requestDevice, always return true in `adapter/device.features.has(featureName)` (but can't be enumerated because Safari doesn't know the name - this is very weird)
  - Limits always start with some string indicating they're part of `compatibility` as well as what their limit class is (`maximum`/`alignment`). Safari specially treats those in `GPUSupportedLimits` such that accessing unknown members with these names returns the least-limiting values (Infinity/1 respectively). requestDevice ignores them. (They still can't be enumerated in `Object.keys` for the same reason.)

**Comment by greggman** [comment]
Throwing out another idea, I don't think I like it and I expect no one else will either but I'm just going to write it down anyway.

What if compat is it's own thing just like WebGL1 and WebGL2 are their own things.

It kind of sucks, like WebGL1 and WebGL2

* In WebGL1 you use `tex2DImage(..., gl.RGBA, ..., gl.RGBA, gl.FLOAT)` to make a rgba32float texture. 
* In WebGL2 you use `tex2DIamge(..., gl.RGBA32F, ...., gl.RGBA, gl.FLOAT)` to make a rgba32float texture. (the code for WebGL1 fails in WebGL2)

--

* In WebGL1 you get `EXT_disjoint_timer_query`
* In WebGL2 you get `EXT_disjoint_timer_query_webgl2`

--

* In WebGL1 you request `OES_vertex_array_object` and the functions are on an extension object
* In WebGL2 vertex arrays just work from the main API and trying to ask for `OES_vertex_array_obejct` will fail.


I'm not saying I like this idea but, @kdashg brought up that requesting core and if it fails then requesting compat was acceptable since it's the same as requesting WebGL2 if it fails then requesting WebGL1.

Having them be separate APIs avoids figuring all of this out. You just have to know that they're different APIs that are 95% similar and that most of your code will work but they aren't the same API and you'll have to deal with it in your code.

I know the goal for compat was *"Every compat program is a valid core program"* and I'm 100% for that goal. I'm just throwing this out there because every WebGL1 program is not a valid WebGL2 program.

The one reason I thought it was worth posting this idea is @mwyrzykowski's concern that adding new limits to compat will break stuff on core. Eailer today I thought we might need a new limit when we add arrays of textures. I don't think we'll need a new limit but it kickstarted this idea. If you add the new limit for compat, a dev uses the new limit, a user uses a browser with only core without the new limit, then it breaks, effectively making them 2 APIs.

I guess, even without compat, new limits are added to core, devs use them, older implementations don't have the new limit, apps break if they are unaware that limit might not exist and their code expects it to. So maybe the fact that adding limits/features to compat after shipping breaks core is just something devs have to deal with. They have to put checks in their code.

**Comment by mwyrzykowski** [comment]
> * We come up with some new mechanism for these features and limits. I'll sketch an example - maybe there is a better design I haven't thought of, but I think it will be really complicated and still be a very incomplete solution.
>   
>   * Features always start with `compatibility-` and are ignored by Safari's requestDevice, always return true in `adapter/device.features.has(featureName)` (but can't be enumerated because Safari doesn't know the name - this is very weird)

This could be a solution to my suggestion of `features and limits which are only needed in compat mode only cause errors if not requested in compat mode.` by, as stated, ignoring ones which start with a prefix.

WebKit is certainly able to add features / limits which are no-ops in core. The concerning issue is the time the feature is added to the time WebKit / Safari is able to add it, since Safari ships features two or maybe three times a year.

While I can not comment on software schedules, a hypothetical example:

* Compat feature is proposed in month `A`, added to the specification after some discussion in `A + 2`. 
* Website add new compat feature in `A + 5`, couldn't have been added earlier since trunk was not taking changes for typical fall release in `A + 2`
* Earliest Safari may likely add support is `A + 8`
* Website either needs to workaround lack of Safari support in compat until December or site is broken in Safari for `3` months, which is unfortunate, because site was trying to be more compatible by adopting compat mode

Other UAs may have less frequent updates than 2-3x per year.

We could certainly do something where we pull a list of features from a trusted source which we can update out of band, but this exposes some potential security risk where a feature is enabled where it should not be, so WebKit would prefer not to head down this route.

Prefixing the features / limits which are only applicable to a certain feature level or below seems reasonable. It could help with websites targeting a certain feature level which ones are necessary to specify.

**Comment by SenorBlanco** [comment]
I'm not sure I understand the risk of new Compat features. The underlying APIs (OpenGL ES and D3D11) are not evolving, so unless we miss something now, it seems unlikely we'd have to add new Compat-specific features once the initial Compat spec is nailed down. That said, any new Core feature which would be limited by Compat should be carefully evaluated so that the spec reflects its implementation in Compat. That shouldn't bring up any version skew issues, since the Core feature and Compat limitations (if any) would land in implementations at the same time.

If it helps, here are the things we've agreed to in the proposal so far which Core implementations would have to implement:

- `maxStorageBuffersInVertexStage`, `maxStorageTexturesInVertexStage` limits (default `max(GPUSize32)` in Core)

Here are additional things we think we'll need:

- `float16-renderable` and `float32-renderable` features
- `maxStorageBuffersInFragmentStage`, `maxStorageTexturesInFragmentStage` limits (default `max(GPUSize32)` in Core)

**Comment by kainino0x** [comment]
> This could be a solution to my suggestion of `features and limits which are only needed in compat mode only cause errors if not requested in compat mode.`. As @kainino0x states, we could ignore ones which start with a prefix, so they would not cause errors.

I want to stress that my sketch is a really incomplete solution. I think it's problematic that it would show up in `.has()` but not when iterated. However maybe there is a version of this that works better. Mulling over it, here's a new rough idea (not proven out at all, may totally not work, and not necessarily better than exhaustively listing all the features now):

- Add to core a way to check for presence of compat-specific features, instead of listing them in `adapter.features`, _without making them enumerable_. We would use this only for capabilities which are subsets of core, so that core adapters can just unconditionally return `true` (in all browsers). In `GPUDeviceDescriptor` we could add a new `requiredCompatibilityFeatures` which can be entirely ignored by core adapters.

> I'm not sure I understand the risk of new Compat features. The underlying APIs (OpenGL ES and D3D11) are not evolving, so unless we miss something now, it seems unlikely we'd have to add new Compat-specific features once the initial Compat spec is nailed down.

Assuming we don't miss anything:
- Limits: as long as we add all of them and don't hardcode any, we shouldn't need to add any later.
- Features: every limitation of compat could potentially be its own compat-specific feature (at least if they exist in ES3.1/3.2/D3D11 (assuming compat-on-vulkan-without-core is not a significant configuration)). If we don't catch every such feature off the bat, then we may find that there's some key feature supportable on some but not all Compat-only devices.

  It does seem possible to design something where we commit not to add new compat-specific features later. We could enumerate all of the limitations of Compat at the finest reasonable granularity. Determining the granularity would require looking at the granularity in WebGL 2, OpenGL ES 3.1, and potentially looking at real devices to determine the granularity of things that aren't extensions (especially for framebuffer completeness and texture format capability type things)

I was going to argue that we should still consider a third option requiring applications to just do things right (always check `adapter.features` before trying to request a feature), and test in Safari to make sure, but:
- testing in Safari is very inaccessible to a lot of developers (we would need another browser to have a core-only mode to test safari-like behavior, and developers would need to know about it).
- I had another idea to _not_ expose these features on core adapters so the only testing needed is to request core, but this makes both interoperability and unshippability worse.

**Comment by SenorBlanco** [comment]
> If we are very certain we won’t need any more than the following for compat:
> 
> > * `maxStorageBuffersInVertexStage`, `maxStorageTexturesInVertexStage` limits (default `max(GPUSize32)` in Core)
> 
> and
> 
> > * `float16-renderable` and `float32-renderable` features
> > * `maxStorageBuffersInFragmentStage`, `maxStorageTexturesInFragmentStage` limits (default `max(GPUSize32)` in Core)
> 
> then I would be fine closing this issue and adding these to core as the list is quite small.

There is one other issue we've discovered, which is that of the combined limits (GL_MAX_COMBINED_SHADER_STORAGE_BLOCKS, GL_MAX_COMBINED_IMAGE_UNIFORMS, GL_MAX_COMBINED_SHADER_OUTPUT_RESOURCES). This may require at most three more limits; see https://github.com/gpuweb/gpuweb/issues/5003

To discover any further issues, I ran the Compat CTS on top of ANGLE with the limits forced to the higher of the GLES spec minimum and the Compat minimum. There were no further adjustments required to pass the CTS other than what's listed above (including the three combined limits). So I'm fairly confident we've discovered them all.

> My concern was primarily with discovering something several months later from now.

Understood. I appreciate the concern.

**Comment by mwyrzykowski** [comment]
This is great information, thank you @SenorBlanco 

In that case, it should be fine to add these features/limits to core and then we don't need to develop some method to make this backwards compatible.

**Comment by kainino0x** [comment]
I think we can be reasonably confident about the limits, but I think figuring out the exact set of features is a bit harder since they might be somewhat arbitrarily broken up.

Also this plan still does require us to discover, design, and standardize all of the limits/features before Safari branches to ship.

If we're willing to forgo the ability to enumerate compat features then I think it's a decent option to keep the compat limits in core but separate out the compat features into a separate `adapter.hasCompatibilityFeature(DOMString) -> boolean` + `GPUDeviceDescriptor.requiredCompatibilityFeatures : sequence<DOMString>`.

Probably if we have a core `everything-from-compat-to-core` feature (as I've described on other issues) which combines all of the un-enumerable features, then enumerating features is not important.

**Comment by kainino0x** [comment]
> Probably if we have a core `everything-from-compat-to-core` feature (as I've described on other issues) which combines all of the un-enumerable features, then enumerating features is not important.

#5036

> If we're willing to forgo the ability to enumerate compat features then I think it's a decent option to keep the compat limits in core but separate out the compat features into a separate `adapter.hasCompatibilityFeature(DOMString) -> boolean` + `GPUDeviceDescriptor.requiredCompatibilityFeatures : sequence<DOMString>`.

For the record there was more discussion of this here: https://github.com/gpuweb/gpuweb/pull/5018#issuecomment-2539929838

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2025-01-22#how-to-add-features-and-limits-for-compat-after-core-ships-4987-includes-hascompatibilityfeaturerequiredcompatibilityfeatures-proposal">GPU Web WG 2025-01-22 Atlantic-time</a></summary>

* KN: summary
* MW: I think this is a better proposal than what we had before
* JB: Favorite because of low impact on core-only browsers?
* MW: That, and not requiring bikeshedding all of the names for compat in the near future, have time to decide names, which features are needed, etc.
* JB: Our first choice is no individuation of features at all, in which case this wouldn't be necessary.
* SW: My pros and cons:
    * Pros: separates compat features into their own namespace; core-only easier; unshipping compat easier
    * Cons: Compat code looks different from core code, and pattern doesn't match
    * Purpose is to protect us during a period where not all browsers have synced with the latest spec
* KG: May find us not wanting to [] in some cases. Imagine we hadn't discovered we couldn't do float32-blending by default. If we add that and put it in the spec and a browser hasn't updated yet, users would start getting warnings 
* KN: Lets us defer work on un-restricting things that we restricted in Compat.
    * Does not help us if we mess up and failed to identify a restriction that should have been in compat
* JB: CTS still in progress right? Further work in CTS may discover new limitations
* JB: The damage is that there will be [some devices that could have had compat, that cannot because compat requires some capability]
* JB: Not a fan of doing a lot of work for every last sliver of [hardware] audience. Not the end of the world if we can't offer them an adapter.
* KN: Can easily imagine features that are say, 50% of compat. Too big to require in compat, too important to eliminate completely in compat.
    * A lot of WebGL extensions' equivalents fall into this category
* JB:
* KG: Multiple stories for compat… one end is go back as far as you can, minimal featureset. Other is supporting devices that just missed core. Has me spitballing in my head, could we just have a very small number of tiers? (Not to send us down an intense rabbithole)
* SW: Mali for silly reasons can't support storage buffers in .. stages.... A lot of overlap in which major features are supported at what point.
* GT: Think the major features (not limits) that will support a lot of devices are float16-renderable and float32-renderable
* SW: e.g. D3D11 will support texture views. Could be supported.
* JB: All of this needs to be phrased in terms of incremental audience. If we *must* start discriminating features in compat, that changes my thinking.

</details>

**Comment by kainino0x** [comment]
I have written up example code for how the various combinations of proposals would work. Apologies it's a bit verbose but hopefully this highlights the differences better:

<https://docs.google.com/document/d/1LWxtRy8dR-T12bxBcBJXsdx0k9km5w9pn72TzwAj04Y/edit?usp=sharing>

**Comment by kainino0x** [comment]
Also I should highlight the specific proposal from Jim in the last meeting to simply _not_ have any Compat features, i.e. you take all of the compat restrictions or none of them.

See https://github.com/gpuweb/gpuweb/issues/4417#issuecomment-2613051118 which talks about removing the two we have so far.

**Comment by kainino0x** [comment]
I wrote up some stuff about future work post-#5036 if we do end up adding features to compat. I will post this here and then link to it from the proposal for reference.

---

Some capabilities which are disallowed in baseline Compatibility mode could be incrementally supported on devices which still do not support Core.

There are four limits which already allow this: `maxStorage{Buffers,Textures}In{Vertex,Fragment}Stage`.

Initially, there are no such features (though Compatibility mode adapters may have features that are defined in core, such as `"float32-blendable"`).
Two designs have been considered for adding such features. They are noted here for possible future use.

- Add them to the existing `adapter.features`/`device.features`/`requiredFeatures`.
  Applications do something like:

  ```js
  adapter.features.has("webgpu-core") || adapter.features.has("int-multisample")
  ```

- Add new `adapter.hasCompatibilityFeature()`/`device.hasCompatibilityFeature()`/`requiredCompatibilityFeatures`, or similar.
  Applications do something like:

  ```
  adapter.hasCompatibilityFeature === undefined || adapter.hasCompatibilityFeature("int-multisample")
  ```
  or
  ```
  adapter.features.has("webgpu-core") || adapter.hasCompatibilityFeature("int-multisample")
  ```

In all 3 sample code lines, the left side of the `||` is important to avoid breaking on browsers that don't expose everything yet.

For earlier comparisons, see the 3rd and 4th columns of [this spreadsheet](https://docs.google.com/document/d/1LWxtRy8dR-T12bxBcBJXsdx0k9km5w9pn72TzwAj04Y/edit?usp=sharing).

**Comment by Kangz** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/GPU-Web-2025-01-29#how-to-add-features-and-limits-for-compat-after-core-ships-4987">GPU Web WG 2025-01-29 Atlantic-time</a></summary>

    * JB: Controversy in the past was that Mozilla preferred to have no features for compat. If some feature were closer to 50% then we would want to expose it. However right now there are no such features proposed. Mozilla would prefer not to take any of this machinery.
    * Kai's doc of code samples: [compat init case studies](https://docs.google.com/document/d/1LWxtRy8dR-T12bxBcBJXsdx0k9km5w9pn72TzwAj04Y/edit)
    * JB: I noticed looking at that, that having a "webgpu-core" feature makes the code look pretty nice. Independent of everything else - having it be an ordinary feature, that everyone would provide, and that you could be auto-upgraded to (not guaranteed to be validated) - but it made the code look nicer. Warm to that proposal.
    * SW: Also in favor of returning adapter.featureLevel?
    * JB: Yes. Agree with Kai's footnote [1] in the doc.
    * GT: I don't see the need for "webgpu-core" in that, back to a year ago, you would just requestAdapter(core) then fall back to requestAdapter(compat).
    * JB: Fine either way, both code samples look OK.
    * BJ: I brought up some concerns last time about it feeling weird and slightly backward. Gave it thought after that, and I didn't think of anything that felt less backward or weird. As Jim pointed out I can see how this allows fairly clean code. Retracting previous "bleh".
    * SW: One thing that became clear with this exercise, if you're willing to have cascading requestAdapter, then we don't need adapter.featureLevel or "webgpu-core".
    * GT: If we add compat features, then if you write code that's compat-only it will fail on core. You'll get upgraded to core without being able to easily see that you were upgraded, and then requesting the compat feature will fail.
    * SW: aside: There are some things in ES3.2 like per-color-target state, that we could expose…
    * KN: Don't think there was any way to totally avoid issues with adding compat features, without having hasCompatibilityFeature.
    * JB: The value of adding the feature depends on how much audience it adds. This audience would be strictly people who require compat but also require this other feature. It seems like a small slice.
    * SW: Have heard that issue with vertex SSBOs but not others yet.
    * GT: And that is solved with limits already.
    * KN: FWIW can add hasCompatibilityFeature in the future. People just have to check it exists before calling it.
    * KG: Should keep in mind how easy things would be to remove. Prefer to 
    * KN: Sounds like current proposal is: no adapter.featureLevel, no hasCompatibilityFeature, no compatibility features at all.
    * GT: No way to tell if it's compat except by which requestAdapter you got the adapter from.
    * KG: Should be differentiable without having to know what you asked for. If I hand you my just my GPUDevice.
    * KN: See "Runtime feature detection examples" footers in the spreadsheet.
    * GT: Benefit of "webgpu-core" feature.
    * KG: We will do this in the future when we move optional features to required.
    * JB: 
    * GT: requesting "webgpu-core" will fail on a device that doesn't have core …
    * JB: 
    * JB: Is it our expectation that in the future some features will become required?
    * KG: Yes, I expect this. Though we haven't done it with WebGL.
    * KN: I also expect this.
    * KG: Features and feature levels are very forward-compatible. If you ask for "compatibility" feature level and get the "webgpu-core" feature, there's no way to ask for compat without core, without some browser flag or something. Thinking that is okay…
    * KN: If you requestAdapter(compat) and then don't enable "webgpu-core", the compat restrictions will be enforced.
    * GT: If you requestDevice w/"webgpu-core", do you also get Core limits?
    * KN: Yes, that's my proposal.
    * JB: Draw up a new PR that clearly explains all the decisions behind it.
    * **KN: So proposal is: No adapter.featureLevel, no compatibility features, yes "webgpu-core"?**
    * JB/MW/KN: thumbs up
    * KG: Maybe “core” instead of “webgpu-core”?
    * KN: OK with me, just picked `"webgpu-core"` so it was a different string than the feature level `"core"`.
    * **PR needed**

</details>

**Comment by Kangz** [comment]
We decided to not add features to compat that are not in core, so this issue is "won't fix".
